autoload/lsp/internal/semantic.vim

```diff
@@ -0,0 +1,216 @@
+    endif
+
+    return lsp#request(l:servers[0], {
+        \ 'method': 'textDocument/semanticTokens/full',
+        \ 'params': {
+        \     'textDocument': lsp#get_text_document_identifier()
+        \ }})
+endfunction
+
+" Highlight helper functions {{{1
+function! s:handle_semantic_request(data) abort
+    call lsp#log('handle_semantic_request', a:data)
+    if lsp#client#is_error(a:data['response'])
+        call lsp#log('Skipping semantic highlight: response is invalid')
+        return
+    endif
+
+    let l:server = a:data['server_name']
+    let l:uri = a:data['request']['params']['textDocument']['uri']
+    let l:path = lsp#utils#uri_to_path(l:uri)
+    let l:bufnr = bufnr(l:path)
+
+    " Skip if the buffer doesn't exist. This might happen when a buffer is
+    " opened and quickly deleted.
+    if !bufloaded(l:bufnr) | return | endif
+
+    call s:init_highlight(l:server, l:bufnr)
+    call s:clear_highlights(l:server, l:bufnr)
+
+    let l:linenr = 0
+    let l:col = 0
+    let l:result_data = a:data['response']['result']['data']
+    while len(l:result_data) > 0
+        let l:linenr = l:linenr + l:result_data[0]
+        if l:result_data[0] > 0
+            let l:col = 0
+        endif
+        let l:col = l:col + l:result_data[1]
+        let l:length = l:result_data[2]
+        let l:token_idx = l:result_data[3]
+        let l:token_modifiers = l:result_data[4]
+
+        call s:add_highlight(l:server, l:bufnr, l:linenr, l:col, l:length, l:token_idx, l:token_modifiers)
+
+        let l:result_data = l:result_data[5:]
+    endwhile
+endfunction
+
+function! s:init_highlight(server, buf) abort
+    if s:use_vim_textprops
+        let l:legend = lsp#internal#semantic#get_legend(a:server)
+        for l:token_idx in range(len(l:legend['tokenTypes']))
+            let l:token_name = l:legend['tokenTypes'][l:token_idx]
+            let l:hl = s:get_hl_name(a:server, l:token_name)
+            let l:textprop_name = s:get_textprop_name(a:server, l:token_idx)
+            silent! call prop_type_add(l:textprop_name, {'bufnr': a:buf, 'highlight': l:hl, 'combine': v:true})
+        endfor
+    endif
+endfunction
+
+function! s:clear_highlights(server, buf) abort
+    if s:use_vim_textprops
+        let l:legend = lsp#internal#semantic#get_legend(a:server)
+        for l:token_idx in range(len(l:legend['tokenTypes']))
+            let l:textprop_name = s:get_textprop_name(a:server, l:token_idx)
+            silent! call prop_remove({'bufnr': a:buf, 'type': l:textprop_name, 'all': v:true}, 1, line('$'))
+        endfor
+    endif
+endfunction
+
+function! s:add_highlight(server, buf, line, col, length, token_idx, token_modifiers) abort
+    let l:legend = lsp#internal#semantic#get_legend(a:server)
+
+    if s:use_vim_textprops
+        try
+            let l:textprop_name = s:get_textprop_name(a:server, a:token_idx)
+            call prop_add(a:line + 1, a:col + 1, { 'length': a:length, 'bufnr': a:buf, 'type': l:textprop_name})
```

Comment:
does +1 work correctly for multibyte chars? do we need to use utils from https://github.com/prabirshrestha/vim-lsp/blob/master/autoload/lsp/utils/position.vim?